Memory leak during internal object listing

williamlardier opened the pull request:
- Happens when the Skip algorithms tells the function to skip an entry.
- In this case, we recursively call the internalListObject function.
- However, we never properly close the stream created from the Transform object, that requires specific cleanup.
- Overall, the logic is too complex, so cleaning is now centralized and modular, so we clean anything that is still there, in a more clear fashion: we clean any of the three streams. If the stream was already destroyed, the operation is a no-op, but at least, all paths now lead to a consistent cleaning of the resources.

Issue: ARSN-437

For reference, some tests were started to automate memory leak detection, but not very stable so far, or require to run it for long times: https://github.com/scality/Arsenal/compare/bugfix/ARSN-437-ballooning?expand=1

Seragonia committed:
Fix memory leak

- Happens when the Skip algorithms tells the function to skip
  an entry.
- In this case, we recursively call the internalListObject
  function.
- However, we never properly close the stream created from
  the Transform object, that requires specific cleanup.
- Overall, the logic is too complex, so cleaning is now
  centralized and modular, so we clean anything that is
  still there.

Issue: ARSN-437

Seragonia committed:
Add tests to cover a listing path with recursive calls

Issue: ARSN-437

Seragonia committed:
Bump to 8.1.139

Issue: ARSN-437

bert-e commented:
# Hello williamlardier,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/arsenal/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ARSN-437 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.1.139`


Please check the `Fix Version/s` of ARSN-437, or the target
branch of this pull request.

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -1795,7 +1810,7 @@ class MongoClientInterface {
                     log.error(
                         'internalListObjectV1: error listing objects', logObj);
-                    baseStream.destroy();
+                    cleanupStreams();
```
it seems we have already a mecanism to handle the propagation of events accross streams and associated cleanup (with a bug indeed):
```
                resolvePhdKey.on('end', () => {
                    baseStream.emit('end');
                });

	[...]

	skip.setListingEndCb(() => {
            stream.emit('end');
            stream.destroy();
        });

	[...]

	    .on('end', () => {
                const data = extension.result();
                // call explicitly the destroy method to clean the mongodb cursor
                stream.destroy();
                cbOnce(null, data);
            });
```

instead of adding a "global" cleanup called multiple times, making it difficult to understand where it will be called and thus if it will always be called exactly once: it seems better to let each transform handle its own cleanup AND propagate the signals/errors.

(i.e. we should preferably just fix the existing "layered" approach, which is consistent with the "stacking streams" approach used here ; and if not possible remove preexisting cleanup code, and do it just once at the end)

williamlardier replied:
I fear "just" fixing the current code will likely let the issue come back someday if we miss another cleanup, which is the mistake we did last time. The current leak was already there, but since we did not see it, we forgot to add cleanup for this case. Given all the code paths here, I am definitely not confident that fixing only the current leak (line 1853) will fix everything for all code paths, unless we clean everything in a consistent way.

I vote for a wrapper at the en (cbOnce), so we clean every time we complete one function call. However I will need to spend time again to confirm this still works, due to the mix of recursive calls, events, and callbacks...

williamlardier replied:
Is it what you had in mind? https://github.com/scality/Arsenal/pull/2264/commits/d2d5391bdcac821aad1d91f6240993ba3cd0af5c

francoisferrand replied:
still some "duplicated" `stream.destroy`:

```
        skip.setListingEndCb(() => {
            stream.emit('end');
            stream.destroy();
        });
        skip.setSkipRangeCb(range => {
            // stop listing this key range
            stream.destroy();
```

(on the second one, maybe we do indeed want to destroy the stream before proceeding with the listing: but then should do the whole cleanup, and reset `stream` to a falsy value)

williamlardier replied:
I'm not sure we should remove this, I need to test first, sending "end" seem more appropriate, to keep the cleanup in a single place

williamlardier replied:
Note: This PR will not be merged till we re-confirm the latest changes are stable on a real platform. Expecting no merge before next monday.

williamlardier replied:
So we can remove the first destroy indeed (but keep the end event). For the second, we indeed must keep the cleanup, otherwise we may get more data from the stream in the parent function, making the whole thing never returning the callback.

- I also updated the logic to only destroy the baseStream if it is not the `stream` object.
- All streams are set to `null` -> not formally needed, but you suggested it and it makes it explicit for the garbage collector that we don't use the variable anymore, so likely to be cleaned faster than before.

https://github.com/scality/Arsenal/pull/2264/commits/8961aa53c1a9b9ee637a5f12376e7ee15550299e

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

Seragonia committed:
Make the internal listing cleanup logic consistent

To avoid missing code paths where some sockets are not properly
cleaned, we now clean each of them when calling cbOnce, to ensure
the logic is called once per function (or recursive) call.

Issue: BB-602

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -1734,9 +1734,24 @@ class MongoClientInterface {
+                baseStream.removeAllListeners();
+                baseStream.destroy();
+            }
+            if (stream) {
```
in some case (Bucket v0 case), `baseStream` may be the same as `stream` : so we should handle this in the callback to avoid using the destroyed stream...

williamlardier replied:
I don't see any issue, as in nodejs, destroying an already destroyed stream is a no-op...

Seragonia committed:
Only destroy streams in cbOnce, and "end" from events

A specific case, before calling recursively the function,
require a cleanup of the resources.

Issue: BB-602

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ARSN-437 contains:


* `8.1.139`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.1.139`

* `8.2.0`


Please check the `Fix Version/s` of ARSN-437, or the target
branch of this pull request.

bert-e commented:
# Request integration branches



Waiting for integration branch creation to be requested by the user.

To request integration branches, please comment on this pull request with the following command:

```
/create_integration_branches
```

Alternatively, the `/approve` and `/create_pull_requests` commands will automatically
create the integration branches.

ghost commented:
/create_integration_branches

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/ARSN-437` into
`development/8.1`
* [w/8.2/bugfix/ARSN-437](https://github.com/scality/arsenal/compare/development/8.2...w/8.2/bugfix/ARSN-437) will be merged into `development/8.2` 



The following branches will **NOT** be impacted:

* `development/6.4`
* `development/7.10`
* `development/7.4`
* `development/7.70`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **create_integration_branches**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_integration_branches**

ghost commented:
/approve

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.1`

* :heavy_check_mark: `development/8.2`



The following branches have **NOT** changed:

* `development/6.4`
* `development/7.10`
* `development/7.4`
* `development/7.70`



Please check the status of the associated issue ARSN-437.

Goodbye williamlardier.




*The following options are set:* **approve, create_integration_branches**
